core/src/window_service.rs

```diff
@@ -570,13 +569,87 @@ mod test {
             &duplicate_slot_sender,
         )
         .unwrap();
-        assert!(blockstore.has_duplicate_shreds_in_slot(duplicate_shred_slot));
+
+        // Make sure the correct duplicate proof was stored
+        let duplicate_proof = blockstore.get_duplicate_slot(duplicate_shred_slot).unwrap();
+        assert_eq!(duplicate_proof.shred1, *original_shred.payload());
+        assert_eq!(duplicate_proof.shred2, *duplicate_shred.payload());
+
+        // Make sure a duplicate signal was sent
         assert_eq!(
             duplicate_slot_receiver.try_recv().unwrap(),
             duplicate_shred_slot
         );
     }
 
+    #[test]
+    fn test_store_duplicate_shreds_same_batch() {
+        let ledger_path = get_tmp_ledger_path_auto_delete!();
+        let blockstore = Arc::new(Blockstore::open(ledger_path.path()).unwrap());
+        let (duplicate_shred_sender, duplicate_shred_receiver) = unbounded();
+        let (duplicate_slot_sender, duplicate_slot_receiver) = unbounded();
+        let exit = Arc::new(AtomicBool::new(false));
+        let keypair = Keypair::new();
+        let contact_info = ContactInfo::new_localhost(&keypair.pubkey(), timestamp());
+        let cluster_info = Arc::new(ClusterInfo::new(
+            contact_info,
+            Arc::new(keypair),
+            SocketAddrSpace::Unspecified,
+        ));
+
+        // Start duplicate thread receiving and inserting duplicates
+        let t_check_duplicate = WindowService::start_check_duplicate_thread(
+            cluster_info,
+            exit.clone(),
+            blockstore.clone(),
+            duplicate_shred_receiver,
+            duplicate_slot_sender,
+        );
+
+        let handle_duplicate = |shred| {
+            let _ = duplicate_shred_sender.send(shred);
+        };
+        let num_trials = 100;
+        for slot in 0..num_trials {
```

Comment:
The bug wasn't deterministic because it's possible the insertion thread finishes *before* the duplicate detection thread receives the shred signal.

After the fix it is now deterministic, but it would be good to catch future bugs like this, hence the multiple trials